### PR TITLE
Add pricing function

### PR DESCRIPTION
**Pull request by davidrpugh**
@bherd I have added traits for discriminatory and uniform pricing rules similar to those described in Wellman et al (1998) and implemented a discriminatory weighted average pricing rule that nests the discriminatory rules discussed in Wellman et al (1998) as special cases.

### CODE REVIEW

File: src/main/scala/org/economicsl/auctions/pricing/DiscriminatoryPricingRule.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package org.economicsl.auctions.pricing
+
+import org.economicsl.auctions.{LimitAskOrder, LimitBidOrder, Price}
+
+
+/** Discriminatory pricing rule computes a different price for each pair of orders. */
+trait DiscriminatoryPricingRule extends (((LimitAskOrder, LimitBidOrder)) => Price)
```
Comment (bherd-rb) [inline src/main/scala/org/economicsl/auctions/pricing/DiscriminatoryPricingRule.scala]: Would be interested to discuss what that exactly means (syntactic-wise)

**Review by bherd-rb — APPROVED** [review]

File: src/main/scala/org/economicsl/auctions/sandbox.sc [hunk 2]
```diff
@@ -30,4 +32,39 @@ val order7: MarketBidOrder = MarketBidOrder(issuer, Quantity(100), tradable)
...
+pairedOrders.toList
+
+
+// Implement a weighted average pricing rule...
+case class WeightedAveragePricing(weight: Double) extends DiscriminatoryPricingRule {
+
+  def apply(pair: (LimitAskOrder, LimitBidOrder)): Price = pair match {
+    case (askOrder, bidOrder) => Price((1 - weight) * askOrder.limit.value + weight * bidOrder.limit.value)
+  }
+
```
Comment (davidrpugh) [inline src/main/scala/org/economicsl/auctions/sandbox.sc]: @bherd-rb I think that the pricing rules should be put into the core library.  In most all cases users would want to implement an off-the shelf pricing rule.  We should try to encourage uses implementing bespoke pricing rules to contribute them back...

**Comment by davidrpugh** [comment]
@bherd-rb This PR is not ready to merge yet.  I have updated to reflect changes from the `develop` branch but more thought needs to go into the type signatures for the pricing rules.

**Comment by davidrpugh** [comment]
@bherd-rb I have added a base trait for describing what I think is the right type signature for a pricing rule.  A pricing rule should map the state of the auction to a price.  The state of the auction is entirely captured by the order book; we also need to deal with the possibility that in some states the price might be undefined.

```scala
trait PricingRule[-T <: Tradable] extends ((FourHeapOrderBook[T] => Option[Price])
```

Thoughts?

**Comment by bherd-rb** [comment]
@davidrpugh I think that looks good. What is your opinion on the return value? Is it always `Price` or could it be some other subtype of `AnyVal`? Could someone be interested in, e.g. valuing data in terms of other data rather than money (no idea if that makes sense...)? 

```scala
trait PricingRule[-T <: Tradable, +V <: AnyVal] extends ((FourHeapOrderBook[T]) => Option[V])
```

**Comment by davidrpugh** [comment]
@bherd-rb This definitely makes sense!  For example in most of the functional smart contract implementations that I have seen, `Currency` is actually a type of `Contract` rather than a primitive `Double`.  I think it is definitely possible for an auction to price goods that are being sold in non-money terms.  For example, one could imagine an auction where oranges are sold in terms of apples.

This kind of thinking may be particularly relevant in a world of blockchain contracts...

**Comment by davidrpugh** [comment]
@bherd-rb I have made substantial progress cleaning up this branch and re-implementing the pricing rules in terms of our new typing and much closer to the language used in the original Wurman et al (1998) paper.

**Review by bherd-rb — APPROVED** [review]

**Pull request merged**